src/preprocessing/particle_packing/signed_distance.jl

```diff
@@ -0,0 +1,273 @@
+    boundary_packing    :: Bool
+
+    function SignedDistanceField(geometry, particle_spacing;
+                                 point_grid=nothing,
+                                 max_signed_distance=4 * particle_spacing,
+                                 use_for_boundary_packing=false)
+        NDIMS = ndims(geometry)
+        ELTYPE = eltype(max_signed_distance)
+
+        sdf_factor = use_for_boundary_packing ? 2 : 1
+
+        nhs = FaceNeighborhoodSearch{NDIMS}(sdf_factor * max_signed_distance)
+
+        initialize!(nhs, geometry)
+
+        if isnothing(point_grid)
+            min_corner = geometry.min_corner .- sdf_factor * max_signed_distance
+            max_corner = geometry.max_corner .+ sdf_factor * max_signed_distance
+
+            n_particles_per_dimension = Tuple(ceil.(Int,
+                                                    (max_corner .- min_corner) ./
+                                                    particle_spacing))
+
+            grid = rectangular_shape_coords(particle_spacing, n_particles_per_dimension,
+                                            min_corner; tlsph=true)
+
+            point_grid = reinterpret(reshape, SVector{NDIMS, ELTYPE}, grid)
+        end
+
+        positions = [SVector(position) for position in point_grid]
+        normals = fill(SVector(ntuple(dim -> Inf, NDIMS)), length(point_grid))
+        distances = fill(Inf, length(point_grid))
+
+        calculate_signed_distances!(positions, distances, normals,
+                                    geometry, sdf_factor, max_signed_distance, nhs)
+
+        return new{NDIMS, ELTYPE}(positions, normals, distances, max_signed_distance,
+                                  use_for_boundary_packing)
+    end
+end
+
+@inline Base.ndims(::SignedDistanceField{NDIMS}) where {NDIMS} = NDIMS
+
+function Base.show(io::IO, system::SignedDistanceField)
+    @nospecialize system # reduce precompilation time
+
+    print(io, "SignedDistanceField{", ndims(system), "}()")
+end
+
+function Base.show(io::IO, ::MIME"text/plain", system::SignedDistanceField)
+    @nospecialize system # reduce precompilation time
+
+    if get(io, :compact, false)
+        show(io, system)
+    else
+        summary_header(io, "SignedDistanceField{$(ndims(system))}")
+        summary_line(io, "#particles", length(system.distances))
+        summary_line(io, "max signed distance", system.max_signed_distance)
+        summary_footer(io)
+    end
+end
+
+function trixi2vtk(signed_distance_field::SignedDistanceField)
+    (; positions, distances, normals) = signed_distance_field
+    positions = stack(signed_distance_field.positions)
+
+    trixi2vtk(positions, signed_distances=distances, normals=normals,
+              filename="signed_distance_field")
+end
+
+function calculate_signed_distances!(positions, distances, normals,
+                                     boundary, sdf_factor, max_signed_distance, nhs)
+    @threaded positions for point in eachindex(positions)
```

Comment:
How performance-relevant is all of this here? Is it only used once during initialization, or is it a bottleneck somewhere? We could probably speed this up by only looping over neighboring cells of non-empty cells.
But that only makes sense if this is not already fast enough.

Reply:
It is only used once during initialization (so far). However, for big geometries it can be performance relevant. That is, the faster the better. 

> We could probably speed this up by only looping over neighboring cells of non-empty cells.

What exactly do you mean? Check if the point position is in a non-empty cell? But what if the face is in the neighboring cell and closest to this point position?

Reply:
Is looping over cells thread supported?
How can I loop over non-empty cells? 
How do I know which point index is in the non-empty cell?

Reply:
Not out of the box. This would require some careful optimization in the NHS itself. If this is not a bottleneck, then it's not worth it.